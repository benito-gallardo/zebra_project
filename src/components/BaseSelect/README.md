# BaseSelect Component

A global select component with default styling.

## How to use

```
import React from 'react';

import BaseSelect from './components/BaseSelect';

...
<BaseSelect onChange={handleChange} options={options} />
```

## Default Props

| Prop Name | Options | Default Value                          | Description                                                                                                                                                                   |
| --------- | ------- | -------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| onChange  | event   | none                                   | Return the event                                                                                                                                                              |
| options   | array   | `[{ title: "Hello", value: 'World' }]` | Passes an array of objects with the `keys` of `title` and `value`. <br /> `title` = Displayed Text <br /> `value` = Populates the `value` attribute on the `options` element. |
