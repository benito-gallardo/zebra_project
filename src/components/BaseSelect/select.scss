@import '../../assets/sass/vars/index';

select {
  border-color: $grey;
  padding: 5px;
  border-radius: 3px;
  color: $grey;
  outline: none;
  font-size: 14px;
}
