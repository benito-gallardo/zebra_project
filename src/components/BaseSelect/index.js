import React from 'react';

import './select.scss'

const Select = ({
  onChange,
  options = [{ title: "Hello", value: 'World' }]
}) => (
    <select data-testid="base-select" onChange={onChange}>
    {options.map(({title, value}, i) => (
      <option data-testid={`base-select-option`} key={`option-${i}`} value={value}>{title}</option>
    ))}
    </select>
);

export default Select;