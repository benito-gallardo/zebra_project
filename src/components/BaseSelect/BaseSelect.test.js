import { render, cleanup, fireEvent  } from '@testing-library/react';
import React from 'react';
import ReactDOM from 'react-dom';

import BaseSelect from './index';

let Select;
let Options;
const optionsArry = [
  { title: 'Title 0', value: 0 },
  { title: 'Title 1', value: 1 }
];

afterEach(cleanup);

describe('The BaseSelect Component', () => { 
  
  it('renders without crashing', () => { 
    const div = document.createElement('div');
    ReactDOM.render(<BaseSelect />, div);
  })

  it('matches snapshot with props', () => {
    const callback = jest.fn();
    const { asFragment } = render(<BaseSelect options={optionsArry} onChange={callback} />);
    expect(asFragment()).toMatchSnapshot();
  })

  it('matches snapshot without props', () => {
    const { asFragment } = render(<BaseSelect />);
    expect(asFragment()).toMatchSnapshot();
  })
  
  describe('When props are provided', () => { 
    
    beforeEach(() => { 
      const { getByTestId, getAllByTestId } = render(<BaseSelect options={optionsArry} />);
      Select = getByTestId('base-select');
      Options = getAllByTestId('base-select-option');
    })

    it('when options prop is provided', () => {
      expect(Select).toContainElement(Options[0]);
      expect(Options[0]).toHaveTextContent('Title 0');
      expect(Options[0]).toHaveAttribute('value','0');
    })
  })

  describe('when props are not provided', () => { 

    beforeEach(() => { 
      const { getByTestId, getAllByTestId } = render(<BaseSelect />);
      Select = getByTestId('base-select');
      Options = getAllByTestId('base-select-option');
    })

    it('when options prop is not provided', () => {
      expect(Select).toContainElement(Options[0]);
      expect(Options[0]).toHaveTextContent('Hello');
      expect(Options[0]).toHaveAttribute('value','World');
    })
  })

  describe('when BaseSelect is clicked', () => { 
    
    const callback = jest.fn();
    
    beforeEach(() => { 
      const { getByTestId, getAllByTestId } = render(<BaseSelect onChange={callback} options={optionsArry} />)
      Select = getByTestId('base-select');
      Options = getAllByTestId('base-select-option');
    })

    it('calls the callback on selection', () => { 
      fireEvent.change(Select, { target: { value: 1 } });
      expect(callback).toHaveBeenCalledTimes(1);
    })

    it('when an options is selected, selected is true', () => { 
      fireEvent.change(Select, { target: { value: 1 } });
      expect(Options[0].selected).toBeFalsy();
      expect(Options[1].selected).toBeTruthy();
    })
  
  })
})