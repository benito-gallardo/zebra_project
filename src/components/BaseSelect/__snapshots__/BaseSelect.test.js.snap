// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The BaseSelect Component matches snapshot with props 1`] = `
<DocumentFragment>
  <select
    data-testid="base-select"
  >
    <option
      data-testid="base-select-option"
      value="0"
    >
      Title 0
    </option>
    <option
      data-testid="base-select-option"
      value="1"
    >
      Title 1
    </option>
  </select>
</DocumentFragment>
`;

exports[`The BaseSelect Component matches snapshot without props 1`] = `
<DocumentFragment>
  <select
    data-testid="base-select"
  >
    <option
      data-testid="base-select-option"
      value="World"
    >
      Hello
    </option>
  </select>
</DocumentFragment>
`;
